File: python/day06.py
#!/usr/bin/env python

with open("../input/06") as f:
    fish = map(int, f.readline().split(","))

# Part 1
for day in range(80):
    nextfish = list()
    for f in fish:
        if f == 0:
            nextfish.append(8)
            nextfish.append(6)
        else:
            nextfish.append(f - 1)
    fish = nextfish
print(f"Part 1: {len(fish)}")

File: rust/day06.rs
use std::fs;
use std::str::FromStr;

fn fishcount(initial: &Vec<i32>, days: i32) -> i64 {
    // Returns the total number of fish after <days> days, given an initial population <initial>.
    let mut fishcounts: [i64; 9] = [0; 9];
    for f in initial {
        fishcounts[*f as usize] += 1;
    }
    for _d in 0..days {
        let spawncount = fishcounts[0];
        for i in 1usize..9 {
            fishcounts[i - 1] = fishcounts[i];
        }
        fishcounts[8] = spawncount;
        fishcounts[6] += spawncount;
    }
    return fishcounts.iter().sum();
}

fn main() {
    let filename = "../input/06";
    let contents = fs::read_to_string(filename).expect("Reading input file failed!");
    let fish = contents
        .trim()
        .split(",")
        .map(|f| i32::from_str(&f).unwrap())
        .collect();

    let part1 = fishcount(&fish, 80);
    let part2 = fishcount(&fish, 256);
    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);
}
